cpp/oneapi/dal/table/detail/csr.hpp

```diff
@@ -0,0 +1,162 @@
+    const void* get_data() const;
+
+    /// Returns the :literal:`column_indices` pointer.
+    const std::int64_t* get_column_indices() const;
+
+    /// Returns the :literal:`row_indices` pointer.
+    const std::int64_t* get_row_indices() const;
+
+private:
+    explicit csr_table(detail::csr_table_iface* impl) : table(impl) {}
+
+    void check_indices(const std::int64_t row_count,
+                       const std::int64_t column_count,
+                       const std::int64_t* row_indices,
+                       const std::int64_t* column_indices,
+                       const csr_indexing indexing) const {
+        using error_msg = dal::detail::error_messages;
+        const std::int64_t min_value = (indexing == csr_indexing::zero_based) ? 0 : 1;
+        const std::int64_t max_value =
+            (indexing == csr_indexing::zero_based) ? column_count - 1 : column_count;
+        const std::int64_t element_count = row_indices[row_count] - row_indices[0];
+        const std::int64_t last_value = row_indices[row_count];
+
+        if (element_count < 0) {
+            throw dal::domain_error(error_msg::row_indices_lt_min_value());
+        }
+
+        for (std::int64_t i = 0; i <= row_count; i++) {
+            if (row_indices[i] < min_value) {
+                throw dal::domain_error(error_msg::row_indices_lt_min_value());
+            }
+            if (row_indices[i] > last_value) {
+                throw dal::domain_error(error_msg::row_indices_gt_max_value());
+            }
+        }
+
+        for (std::int64_t i = 0; i < element_count; i++) {
+            if (column_indices[i] < min_value) {
+                throw dal::domain_error(error_msg::column_indices_lt_min_value());
+            }
+            if (column_indices[i] > max_value) {
+                throw dal::domain_error(error_msg::column_indices_gt_max_value());
+            }
+        }
+    }
+
+    template <typename Data>
+    void init_impl(const array<Data>& data,
+                   const array<std::int64_t>& column_indices,
+                   const array<std::int64_t>& row_indices,
+                   std::int64_t row_count,
+                   std::int64_t column_count,
+                   csr_indexing indexing) {
+        using error_msg = dal::detail::error_messages;
+
+        if (row_count <= 0) {
+            throw dal::domain_error(error_msg::rc_leq_zero());
+        }
+
+        if (column_count <= 0) {
+            throw dal::domain_error(error_msg::cc_leq_zero());
+        }
+
+        if (indexing != csr_indexing::one_based) {
+            throw dal::domain_error(detail::error_messages::zero_based_indexing_is_not_supported());
+        }
+
+        check_indices(row_count,
+                      column_count,
+                      row_indices.get_data(),
+                      column_indices.get_data(),
+                      indexing);
+
+        table::init_impl(new backend::csr_table_impl(column_count,
```

Comment:
You are not allowed to expose `backend` to the user level. Please hide this in cpp files.

Reply:
Ok, it was fixed